Ezopt command line options and multisend

nrnhines opened the pull request.

pramodk commented:
@nrnhines : I have rebased this PR to latest master. I have also added tests for testing `permutations`, `binqueue`, `multisend` etc.

Instead of create new pull request, I can update this PR by pushing to your ezopt branch.

- Would it be ok if I update your `ezopt` branch?
- Could you check if you have clicked `Allow edits from maintainers` checkbox as shown [here](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/)?

nrnhines commented:
yes for the update.
I'm having trouble with verifying the "Allow edits from maintainers".
Is this done in the BlueBrain/CoreNeuron? But there is no ezopt branch
there. However "Allow edits from maintainers" is checked for a pull request
that I made from nrnhines/CoreNeuron for the ezopt branch.
________________________________________
From: Pramod Kumbhar <notifications@github.com>
Sent: Thursday, July 6, 2017 3:41 PM
To: BlueBrain/CoreNeuron
Cc: Hines, Michael; Mention
Subject: Re: [BlueBrain/CoreNeuron] Ezopt command line options and multisend (#15)

@nrnhines<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_nrnhines&d=DwMFaQ&c=cjytLXgP8ixuoHflwc-poQ&r=RYxvh9hpXg4aGcXdc7VKN3LhQcIm2Bs9VRtAs_7qCcE&m=mpT72pcqORJAmjIkHlChwnNqNNyLoavr8esRRCACriM&s=tHEuz2psoJGts2qXaF3Y6eSdU4BvJ-7oxG4ZjPl9nVg&e=> : I have rebased this PR to latest master. I have also added tests for testing permutations, binqueue, multisend etc.

Instead of create new pull request, I can update this PR by pushing to your ezopt branch.

  *   Would it be ok if I update your ezopt branch?
  *   Could you check if you have clicked Allow edits from maintainers checkbox as shown here<https://urldefense.proofpoint.com/v2/url?u=https-3A__help.github.com_articles_allowing-2Dchanges-2Dto-2Da-2Dpull-2Drequest-2Dbranch-2Dcreated-2Dfrom-2Da-2Dfork_&d=DwMFaQ&c=cjytLXgP8ixuoHflwc-poQ&r=RYxvh9hpXg4aGcXdc7VKN3LhQcIm2Bs9VRtAs_7qCcE&m=mpT72pcqORJAmjIkHlChwnNqNNyLoavr8esRRCACriM&s=tARULPfG9pCaa0nZm1EcKtn2uhhKzBdA8EdSVp6ufcM&e=>?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_BlueBrain_CoreNeuron_pull_15-23issuecomment-2D313498655&d=DwMFaQ&c=cjytLXgP8ixuoHflwc-poQ&r=RYxvh9hpXg4aGcXdc7VKN3LhQcIm2Bs9VRtAs_7qCcE&m=mpT72pcqORJAmjIkHlChwnNqNNyLoavr8esRRCACriM&s=K-GsFpoiXb4Ep52FJiFROfoCKK0t_b3fOv68R8CKjs4&e=>, or mute the thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AEMS0RsVo0p0bu4guVDMHZIBP4Qg70GPks5sLTiCgaJpZM4NND1Q&d=DwMFaQ&c=cjytLXgP8ixuoHflwc-poQ&r=RYxvh9hpXg4aGcXdc7VKN3LhQcIm2Bs9VRtAs_7qCcE&m=mpT72pcqORJAmjIkHlChwnNqNNyLoavr8esRRCACriM&s=pXwdXPEz_XEmapcXum7qnFGMrWGyWnwSxzEm1yD79C0&e=>.

pramodk commented:
>Is this done in the BlueBrain/CoreNeuron?

No, This is usually from the branch where you create pull request from.  

I did rebase and hence I had to use `git push -f`. This allowed me to push to your branch. Now this PR is updated.


I have question about argument. I will ask that on review.

pramodk commented on README.md:
```diff
@@ -135,42 +135,72 @@ In order to see the command line options, you can use:
+                         argument must be specified.
+-o, --outpath ARG        Path to place output data files. (.)
+-p, --pattern ARG        Apply patternstim using the specified spike file.
+-R, --cell-permute ARG   Cell permutation, 0 No; 1 optimise node adjacency; 2
```
Do we want cell permute enabled on CPU?

nrnhines replied:
cell_permute 0 and 1 work on cpu and gpu Is useful for performance comparison and testing.
Cell permute 2 works only for gpu.  On gpu 1 is much faster than 0 and 2 is best. So
a default of 1 is reasonable.

pramodk commented on coreneuron/nrniv/main1.cpp:
```diff
@@ -87,103 +86,111 @@ void nrn_init_and_load_data(int argc, char** argv, cn_input_params& input_params
-    if (input_params.dt != -1000.) {  // command line arg highest precedence
-        dt = input_params.dt;
-    } else if (dt == -1000.) {  // not on command line and no celsius in globals.dat
+    if (nrnopt_get_dbl("--dt") != -1000.) {  // command line arg highest precedence
```
Could you remind me why we check for these special values for  `dt` and `celsius`? (`-1000`)

nrnhines replied:
dt and celsius are normally specified by the globals.dat file. If the user changes away from
the starting sentinal values of -1000. then those supersede the globals.dat values.

pramodk commented on coreneuron/nrniv/nrnoptarg.cpp:
```diff
@@ -29,306 +29,363 @@ THE POSSIBILITY OF SUCH DAMAGE.
+    {"--spikebuf -b", 100000, 1, 2000000000, "Spike buffer size. (100000)"},
+    {"--spkcompress", 0, 0, 100000, "Spike compression. Up to ARG are exchanged during MPI_Allgather. (0)"},
+    {"--prcellgid -g", -1, -1, 2000000000, "Output prcellstate information for the gid NUMBER."},
+    {"--cell-permute -R", 1, 0, 3,
```
enable permutation on CPU? Is there any performance penalty for CPU?

nrnhines replied:
Good question. I'm not sure there is. But was useful for testing.

pramodk commented on coreneuron/nrniv/nrnoptarg.cpp:
```diff
@@ -29,306 +29,363 @@ THE POSSIBILITY OF SUCH DAMAGE.
+    {"--prcellgid -g", -1, -1, 2000000000, "Output prcellstate information for the gid NUMBER."},
+    {"--cell-permute -R", 1, 0, 3,
+     "Cell permutation, 0 No; 1 optimise node adjacency; 2 optimize parent adjacency. (1)"},
+    {"--nwarp -W", 0, 0, 1000000, "Number of warps to balance. (0)"},
```
For GPU, what should be default value for `nwarp`?

nrnhines replied:
0 (unused) is a good default. It takes some experimenting with balance to determine the optimal
value on gpu.

pramodk commented on coreneuron/nrnmpi/nrnmpiuse.h:
```diff
@@ -35,6 +35,11 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #endif
 
+/* define to 1 if want multisend spike exchange available */
+#ifndef NRN_MULTISEND
```
We have command line option to enable/disable multisend. In which case we want to disable compilation of multisend? Should I add that as CMake option? (if require)

nrnhines replied:
If one knows that multisend will not be used, then it is reasonable to disable it. Otherwise
it can be useful to compare performance. It may be very much faster, especially if there are
few connections per cell or number of connections per cell << number of ranks.

pramodk commented on tests/integration/CMakeLists.txt:
```diff
@@ -25,21 +25,39 @@
+    "ring_multisend!${RING_COMMON_ARGS} --outpath ${CMAKE_CURRENT_BINARY_DIR}/ring_multisend --multisend"
+    "ring_permute1!${RING_COMMON_ARGS} --outpath ${CMAKE_CURRENT_BINARY_DIR}/ring_permute1 ${PERMUTE1_ARGS}"
+    "ring_permute2!${RING_COMMON_ARGS} --outpath ${CMAKE_CURRENT_BINARY_DIR}/ring_permute2 ${PERMUTE2_ARGS}"
 
```
This is how I am running/testing ring test with different options (see the arguments).
If any arguments are unnecessary or non-compatible, let me know

nrnhines replied:
looks good to me
